src/basic/recurse-dir.c

```diff
@@ -84,6 +85,12 @@ int readdir_all(int dir_fd,
                 if (ignore_dirent(entry, flags))
                         continue;
 
+                if (FLAGS_SET(flags, RECURSE_DIR_ENSURE_TYPE)) {
```

Comment:
This should be moved to the second loop below, then we can drop entries already removed, that is, `-ENOENT`.